python/ray/autoscaler/docker.py

```diff
@@ -50,31 +65,38 @@ def dockerize_if_needed(config):
                                             docker_mounts, cname,
                                             run_options + worker_run_options)
 
-    config["head_setup_commands"] = head_docker_start + (with_docker_exec(
-        config["head_setup_commands"], container_name=cname))
-    config["head_start_ray_commands"] = (
-        docker_autoscaler_setup(cname) + with_docker_exec(
-            config["head_start_ray_commands"], container_name=cname))
+    config["node_restart_commands"]["head"] = (
+        head_docker_start + config["node_restart_commands"]["head"])
+    config["ray_restart_commands"]["head"] = (
+        docker_autoscaler_setup(cname) +
+        config["node_restart_commands"]["head"])
 
-    config["worker_setup_commands"] = worker_docker_start + (with_docker_exec(
-        config["worker_setup_commands"], container_name=cname))
-    config["worker_start_ray_commands"] = with_docker_exec(
-        config["worker_start_ray_commands"],
-        container_name=cname,
-        env_vars=["RAY_HEAD_IP"])
+    config["node_restart_commands"]["worker"] = (
+        worker_docker_start + config["node_restart_commands"]["worker"])
 
     return config
 
 
-def with_docker_exec(cmds, container_name, env_vars=None):
+def maybe_docker_exec(cmds, container_name, env_vars=None):
+    """Wrap any commands starting with IN_DOCKER in docker exec."""
     env_str = ""
     if env_vars:
         env_str = " ".join(
             ["-e {env}=${env}".format(env=env) for env in env_vars])
-    return [
-        "docker exec {} {} /bin/sh -c {} ".format(env_str, container_name,
-                                                  quote(cmd)) for cmd in cmds
-    ]
+
+    res = []
+    for cmd in cmds:
+        args = cmd.split(" ")
+        if not args[0] == "IN_DOCKER":
```

Comment:
this is really ugly and I don't like the idea of having magic keywords like this in the commands in general. what is the use case for this? why wouldn't these commands either be run in the dockerfile or as part of the application?

Reply:
When using Docker with Ray, we need to run most commands inside the Docker container; e.g. `ray start`.

Currently this wrapping happens implicitly for (what was) `*_setup_commands` and `*_start_ray_commands`. This is problematic: some commands need to run outside of Docker (like mounting EFS instances on AWS), some commands need to run inside Docker (like updating source code of your application). Previously one could use `*_initialization_commands` to run things outside Docker. However, with the recently added support for caching stopped nodes this broke as `initialization_commands` will not get run on node restart.

See https://github.com/ray-project/ray/issues/6128 for discussion that led to this API.

Reply:
The `ray start` command should always be run in the container when using docker so we don't need this syntax for that case.

Seems like if you're using docker the "right" way to do this is to push a new image, but I can see how that would be annoying/high overhead.

Would this be fixed by running always running `node_start_commands` inside of docker? Or do you think there are cases where people need to run commands on both the host and in the container on restart?

Reply:
I think sometimes people need to run commands on the host on restart. For example, mounting external file systems is awkward to do inside of Docker (EC2 credentials may not automatically be passed through).

Good point about the canonical Docker approach being to update source code by pushing a new image. I think this would also require change in Ray's design, though, since right now the Docker container persists until you reboot the node.

I'd be OK removing the `ray start` commands and having a separate config, but think there should be some functionality of running commands each time `ray` is restarted.